File: src/util/main/sp/it/util/type/ObjectFieldMap.java
package sp.it.util.type;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import org.jetbrains.annotations.NotNull;
import sp.it.util.access.fieldvalue.ObjectField;
import sp.it.util.collections.map.ClassMap;
import static java.util.stream.Collectors.toSet;
import static sp.it.util.dev.FailKt.noNull;
import static sp.it.util.functional.Util.stream;

/**
 * Map storing sets of attributes for classes, formally: set of {@link sp.it.util.access.fieldvalue.ObjectField}.
 */
public class ObjectFieldMap {

	private final ClassMap<Set> fields = new ClassMap<>();
	private final ClassMap<Set> cache = new ClassMap<>();
	private final ClassMap<Set> cache2 = new ClassMap<>();

	public <T> void add(Class<T> c, Collection<? extends ObjectField<T,?>> fields) {
		noNull(c);
		noNull(fields);
		fields.forEach(field -> add(c, field));
	}

	@SafeVarargs
	public final <T> void add(Class<T> c, ObjectField<T,?>... fields) {
		noNull(c);
		noNull(fields);
		stream(fields).forEach(field -> add(c, field));
	}

	@SuppressWarnings("unchecked")
	public <T> void add(Class<T> c, ObjectField<T,?> field) {
		noNull(c);
		noNull(fields);
		fields.computeIfAbsent(c, key -> new HashSet<>()).add(field);
		cache.keySet().removeIf(c::isAssignableFrom);
		cache2.keySet().removeIf(c::isAssignableFrom);
	}

	@SuppressWarnings({"unchecked", "RedundantCast"})
	public <T> Set<ObjectField<T,?>> get(Class<T> c) {
		noNull(c);
		return (Set) cache.computeIfAbsent(c, key -> (Set)
			fields.getElementsOfSuperV(key).stream().flatMap(Set::stream).collect(toSet()));
	}

	@SuppressWarnings({"unchecked", "RedundantCast"})
	public <T> Set<ObjectField<T,?>> getExact(Class<T> c) {
		noNull(c);
		return (Set) cache2.computeIfAbsent(c, key -> (Set)
			fields.getElementsOf(key).stream().flatMap(Set::stream).collect(toSet()));
	}

	public static final @NotNull ObjectFieldMap DEFAULT = new ObjectFieldMap();
}

File: src/util/main/sp/it/util/type/ObjectFieldMap.kt
package sp.it.util.type

import kotlin.reflect.KClass
import kotlin.reflect.full.isSuperclassOf
import sp.it.util.access.fieldvalue.ObjectField
import sp.it.util.collections.map.KClassMap
import sp.it.util.functional.asIs

private typealias Field<T> = ObjectField<T, *>

/** Map storing sets of attributes for classes, formally: [Set] of [ObjectField]. */
class ObjectFieldMap {

   private val fields = KClassMap<LinkedHashSet<Field<*>>>()
   private val cache = KClassMap<LinkedHashSet<Field<*>>>()
   private val cache2 = KClassMap<LinkedHashSet<Field<*>>>()

   fun <T: Any> add(c: KClass<T>, fields: Collection<Field<T>>): Unit = fields.forEach { add(c, it) }

   fun <T: Any> add(c: KClass<T>, vararg fields: Field<T>): Unit = fields.forEach { add(c, it) }

   fun <T: Any> add(c: KClass<T>, field: Field<T>) {
      fields.computeIfAbsent(c) { LinkedHashSet() } += field
      cache.keys.removeIf { c.isSuperclassOf(it) }
      cache2.keys.removeIf { c.isSuperclassOf(it) }
   }

   operator fun <T: Any> get(c: VType<T>): Set<Field<T>> = get(c.raw)

   operator fun <T: Any> get(c: KClass<T>): Set<Field<T>> = cache.computeIfAbsent(c) { fields.getElementsOfSuperV(c).flatten().toCollection(LinkedHashSet()) }.asIs()

   fun <T: Any> getExact(c: KClass<T>): Set<Field<T>> = cache2.computeIfAbsent(c) { fields.getElementsOf(c).flatten().toCollection(LinkedHashSet()) }.asIs()

   companion object {
      val DEFAULT = ObjectFieldMap()
   }

}